diff --git a/libc/Android.mk b/libc/Android.mk
index f7f2adc..6e25bd8 100644
--- a/libc/Android.mk
+++ b/libc/Android.mk
@@ -1115,6 +1115,15 @@ LOCAL_NATIVE_COVERAGE := $(bionic_coverage)
 include $(BUILD_STATIC_LIBRARY)
 
 # ========================================================
+# libsprd_jemalloc.a
+# ========================================================
+ifeq ($(TARGET_USE_SPRD_JEMALLOC), true)
+include $(CLEAR_VARS)
+LOCAL_PREBUILT_LIBS := libsprd_jemalloc.a
+include $(BUILD_MULTI_PREBUILT)
+endif
+
+# ========================================================
 # libc_ndk.a
 # Compatibility library for the NDK. This library contains
 # all the parts of libc that are safe to statically link.
@@ -1170,8 +1179,12 @@ LOCAL_WHOLE_STATIC_LIBRARIES_arm := libc_aeabi
 LOCAL_CXX_STL := none
 
 ifneq ($(MALLOC_IMPL),dlmalloc)
+ifeq ($(TARGET_USE_SPRD_JEMALLOC), true)
+LOCAL_WHOLE_STATIC_LIBRARIES += libsprd_jemalloc
+else
 LOCAL_WHOLE_STATIC_LIBRARIES += libjemalloc
 endif
+endif
 
 $(eval $(call patch-up-arch-specific-flags,LOCAL_CFLAGS,libc_common_cflags))
 $(eval $(call patch-up-arch-specific-flags,LOCAL_SRC_FILES,libc_common_src_files))
@@ -1308,8 +1321,12 @@ LOCAL_ADDITIONAL_DEPENDENCIES := $(libc_common_additional_dependencies)
 LOCAL_WHOLE_STATIC_LIBRARIES := libc_common
 
 ifneq ($(MALLOC_IMPL),dlmalloc)
+ifeq ($(TARGET_USE_SPRD_JEMALLOC), true)
+LOCAL_WHOLE_STATIC_LIBRARIES += libsprd_jemalloc
+else
 LOCAL_WHOLE_STATIC_LIBRARIES += libjemalloc
 endif
+endif
 
 LOCAL_CXX_STL := none
 LOCAL_SYSTEM_SHARED_LIBRARIES :=
@@ -1366,8 +1383,12 @@ LOCAL_SHARED_LIBRARIES := libdl
 LOCAL_WHOLE_STATIC_LIBRARIES := libc_common
 
 ifneq ($(MALLOC_IMPL),dlmalloc)
+ifeq ($(TARGET_USE_SPRD_JEMALLOC), true)
+LOCAL_WHOLE_STATIC_LIBRARIES += libsprd_jemalloc
+else
 LOCAL_WHOLE_STATIC_LIBRARIES += libjemalloc
 endif
+endif
 
 LOCAL_CXX_STL := none
 LOCAL_SYSTEM_SHARED_LIBRARIES :=
diff --git a/libc/kernel/uapi/linux/fb.h b/libc/kernel/uapi/linux/fb.h
index 39bb3f4..e260abe 100644
--- a/libc/kernel/uapi/linux/fb.h
+++ b/libc/kernel/uapi/linux/fb.h
@@ -215,6 +215,7 @@ struct fb_bitfield {
 #define FB_ACTIVATE_NOW 0
 #define FB_ACTIVATE_NXTOPEN 1
 #define FB_ACTIVATE_TEST 2
+#define FB_ACTIVATE_NODISP 4
 /* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
 #define FB_ACTIVATE_MASK 15
 #define FB_ACTIVATE_VBL 16
diff --git a/libc/kernel/uapi/linux/ion.h b/libc/kernel/uapi/linux/ion.h
index 4193763..008b03c 100644
--- a/libc/kernel/uapi/linux/ion.h
+++ b/libc/kernel/uapi/linux/ion.h
@@ -72,6 +72,7 @@ struct ion_custom_data {
 #define ION_IOC_SHARE _IOWR(ION_IOC_MAGIC, 4, struct ion_fd_data)
 /* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
 #define ION_IOC_IMPORT _IOWR(ION_IOC_MAGIC, 5, struct ion_fd_data)
+#define ION_IOC_INVALIDATE _IOWR(ION_IOC_MAGIC, 8, struct ion_fd_data)
 #define ION_IOC_SYNC _IOWR(ION_IOC_MAGIC, 7, struct ion_fd_data)
 #define ION_IOC_CUSTOM _IOWR(ION_IOC_MAGIC, 6, struct ion_custom_data)
 #endif
diff --git a/libc/kernel/uapi/video/sprd_adf.h b/libc/kernel/uapi/video/sprd_adf.h
new file mode 100644
index 0000000..a06c68b
--- /dev/null
+++ b/libc/kernel/uapi/video/sprd_adf.h
@@ -0,0 +1,150 @@
+/****************************************************************************
+ ****************************************************************************
+ ***
+ ***   This header was automatically generated from a Linux kernel header
+ ***   of the same name, to make information necessary for userspace to
+ ***   call into the kernel available to libc.  It contains only constants,
+ ***   structures, and macros generated from the original header, and thus,
+ ***   contains no copyrightable information.
+ ***
+ ***   To edit the content of this header, modify the corresponding
+ ***   source file (e.g. under external/kernel-headers/original/) then
+ ***   run bionic/libc/kernel/tools/update_all.py
+ ***
+ ***   Any manual change here will be lost the next time this script will
+ ***   be run. You've been warned!
+ ***
+ ****************************************************************************
+ ****************************************************************************/
+#ifndef _UAPI_VIDEO_SPRD_ADF_H_
+#define _UAPI_VIDEO_SPRD_ADF_H_
+#include <linux/types.h>
+#include <video/adf.h>
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+#define SPRD_ADF_MAX_PLANE 4
+enum {
+ ADF_TRANSFORM_NONE = 0x00,
+ ADF_TRANSFORM_FLIP_H = 0x01,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_TRANSFORM_FLIP_V = 0x02,
+ ADF_TRANSFORM_ROT_90 = 0x04,
+ ADF_TRANSFORM_ROT_180 = 0x03,
+ ADF_TRANSFORM_ROT_270 = 0x07,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_TRANSFORM_RESERVED = 0x08,
+};
+enum {
+ ADF_SCALE_NONE = 0,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+};
+enum {
+ ADF_UNCOMPRESSED,
+ ADF_COMPRESSED,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+};
+enum {
+ ADF_BLENDING_NONE = 0x0100,
+ ADF_BLENDING_PREMULT = 0x0105,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_BLENDING_COVERAGE = 0x0405
+};
+enum {
+ ADF_PIXEL_FORMAT_RGBA_8888 = 1,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_PIXEL_FORMAT_RGBX_8888 = 2,
+ ADF_PIXEL_FORMAT_RGB_888 = 3,
+ ADF_PIXEL_FORMAT_RGB_565 = 4,
+ ADF_PIXEL_FORMAT_BGRA_8888 = 5,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_PIXEL_FORMAT_sRGB_A_8888 = 0xC,
+ ADF_PIXEL_FORMAT_sRGB_X_8888 = 0xD,
+ ADF_PIXEL_FORMAT_YV12 = 0x32315659,
+ ADF_PIXEL_FORMAT_Y8 = 0x20203859,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_PIXEL_FORMAT_Y16 = 0x20363159,
+ ADF_PIXEL_FORMAT_RAW16 = 0x20,
+ ADF_PIXEL_FORMAT_RAW_SENSOR = 0x20,
+ ADF_PIXEL_FORMAT_RAW10 = 0x25,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_PIXEL_FORMAT_RAW_OPAQUE = 0x24,
+ ADF_PIXEL_FORMAT_BLOB = 0x21,
+ ADF_PIXEL_FORMAT_IMPLEMENTATION_DEFINED = 0x22,
+ ADF_PIXEL_FORMAT_YCbCr_420_888 = 0x23,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_PIXEL_FORMAT_YCbCr_422_SP = 0x10,
+ ADF_PIXEL_FORMAT_YCrCb_420_SP = 0x11,
+ ADF_PIXEL_FORMAT_YCbCr_422_P = 0x12,
+ ADF_PIXEL_FORMAT_YCbCr_420_P = 0x13,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_PIXEL_FORMAT_YCbCr_422_I = 0x14,
+ ADF_PIXEL_FORMAT_YCbCr_420_I = 0x15,
+ ADF_PIXEL_FORMAT_CbYCrY_422_I = 0x16,
+ ADF_PIXEL_FORMAT_CbYCrY_420_I = 0x17,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_PIXEL_FORMAT_YCbCr_420_SP_TILED = 0x18,
+ ADF_PIXEL_FORMAT_YCbCr_420_SP = 0x19,
+ ADF_PIXEL_FORMAT_YCrCb_420_SP_TILED = 0x1A,
+ ADF_PIXEL_FORMAT_YCrCb_422_SP = 0x1B,
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ ADF_PIXEL_FORMAT_YCrCb_420_P = 0x1C,
+};
+struct sprd_adf_device_capability {
+ __u32 device_id;
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+};
+struct sprd_adf_interface_capability {
+ __u32 interface_id;
+ __u32 fb_count;
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ __u32 fb_format;
+};
+struct sprd_adf_hwlayer_capability {
+ __u32 hwlayer_id;
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ __u32 format;
+ __u32 rotation;
+ __u32 scale;
+ __u32 blending;
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+};
+struct sprd_adf_overlayengine_capability {
+ __u32 number_hwlayer;
+ __u32 format;
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ __u32 rotation;
+ __u32 scale;
+ __u32 blending;
+ union {
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ struct sprd_adf_hwlayer_capability hwlayers[0];
+ const struct sprd_adf_hwlayer_capability
+ *hwlayer_ptr[SPRD_ADF_MAX_PLANE];
+ };
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+};
+struct sprd_adf_hwlayer_custom_data {
+ __u32 interface_id;
+ __u32 hwlayer_id;
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ __u32 buffer_id;
+ __u32 alpha;
+ __s16 dst_x;
+ __s16 dst_y;
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ __u16 dst_w;
+ __u16 dst_h;
+ __u32 blending;
+ __u32 rotation;
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ __u32 scale;
+ __u32 compression;
+};
+struct sprd_adf_post_custom_data {
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+ __u32 version;
+ __u32 num_interfaces;
+ __s32 retire_fence;
+ struct sprd_adf_hwlayer_custom_data hwlayers[0];
+/* WARNING: DO NOT EDIT, AUTO-GENERATED CODE - SEE TOP FOR INSTRUCTIONS */
+};
+#endif
