diff --git a/minui/Android.mk b/minui/Android.mk
index a8d0aa4..89a4254 100644
--- a/minui/Android.mk
+++ b/minui/Android.mk
@@ -34,6 +34,10 @@ ifeq ($(subst ",,$(TARGET_RECOVERY_PIXEL_FORMAT)),BGRA_8888)
   LOCAL_CFLAGS += -DRECOVERY_BGRA
 endif
 
+ifeq ($(RECOVERY_GRAPHICS_FORCE_SINGLE_BUFFER), true)
+LOCAL_CFLAGS += -DRECOVERY_GRAPHICS_FORCE_SINGLE_BUFFER
+endif
+
 ifneq ($(TARGET_RECOVERY_OVERSCAN_PERCENT),)
   LOCAL_CFLAGS += -DOVERSCAN_PERCENT=$(TARGET_RECOVERY_OVERSCAN_PERCENT)
 else
diff --git a/minui/graphics_fbdev.cpp b/minui/graphics_fbdev.cpp
index b017ff2..1f1a91f 100644
--- a/minui/graphics_fbdev.cpp
+++ b/minui/graphics_fbdev.cpp
@@ -143,6 +143,7 @@ static GRSurface* fbdev_init(minui_backend* backend) {
     memset(gr_framebuffer[0].data, 0, gr_framebuffer[0].height * gr_framebuffer[0].row_bytes);
 
     /* check if we can use double buffering */
+#ifndef RECOVERY_GRAPHICS_FORCE_SINGLE_BUFFER
     if (vi.yres * fi.line_length * 2 <= fi.smem_len) {
         double_buffered = true;
 
@@ -153,6 +154,9 @@ static GRSurface* fbdev_init(minui_backend* backend) {
         gr_draw = gr_framebuffer+1;
 
     } else {
+#else
+    {
+#endif
         double_buffered = false;
 
         // Without double-buffering, we allocate RAM for a buffer to
