diff --git a/include/hardware/gralloc.h b/include/hardware/gralloc.h
index 07ac029..e10ca6a 100644
--- a/include/hardware/gralloc.h
+++ b/include/hardware/gralloc.h
@@ -143,6 +143,7 @@ enum {
     GRALLOC_USAGE_PRIVATE_2             = 0x40000000,
     GRALLOC_USAGE_PRIVATE_3             = 0x80000000,
     GRALLOC_USAGE_PRIVATE_MASK          = 0xF0000000,
+    GRALLOC_USAGE_HW_TILE_ALIGN         = 0x08000000,
 
 #ifdef EXYNOS4_ENHANCEMENTS
     /* SAMSUNG */
diff --git a/include/hardware/hwcomposer.h b/include/hardware/hwcomposer.h
index aa466b3..9ce743e 100644
--- a/include/hardware/hwcomposer.h
+++ b/include/hardware/hwcomposer.h
@@ -808,6 +808,12 @@ typedef struct hwc_composer_device_1 {
     int (*setCursorPositionAsync)(struct hwc_composer_device_1 *dev, int disp, int x_pos, int y_pos);
 
     /*
+     *  Sprd add here:
+     *  Get builtin physical display number from vendor.
+     * */
+    int (*getBuiltInDisplayNum)(struct hwc_composer_device_1 *dev, uint32_t *number);
+
+    /*
      * Reserved for future use. Must be NULL.
      */
     void* reserved_proc[1];
