diff --git a/include/hardware/gralloc.h b/include/hardware/gralloc.h
index 5dafea0..731760c 100644
--- a/include/hardware/gralloc.h
+++ b/include/hardware/gralloc.h
@@ -142,6 +142,7 @@ enum {
     GRALLOC_USAGE_PRIVATE_2             = 0x40000000U,
     GRALLOC_USAGE_PRIVATE_3             = 0x80000000U,
     GRALLOC_USAGE_PRIVATE_MASK          = 0xF0000000U,
+    GRALLOC_USAGE_HW_TILE_ALIGN         = 0x08000000U,
 };
 
 /*****************************************************************************/
diff --git a/include/hardware/hwcomposer.h b/include/hardware/hwcomposer.h
index 9eb1aaf..0e495d6 100644
--- a/include/hardware/hwcomposer.h
+++ b/include/hardware/hwcomposer.h
@@ -773,6 +773,12 @@ typedef struct hwc_composer_device_1 {
     int (*setCursorPositionAsync)(struct hwc_composer_device_1 *dev, int disp, int x_pos, int y_pos);
 
     /*
+     *  Sprd add here:
+     *  Get builtin physical display number from vendor.
+     * */
+    int (*getBuiltInDisplayNum)(struct hwc_composer_device_1 *dev, uint32_t *number);
+
+    /*
      * Reserved for future use. Must be NULL.
      */
     void* reserved_proc[1];
