diff --git a/current.txt b/current.txt
index a476f51..cb6d5fd 100644
--- a/current.txt
+++ b/current.txt
@@ -76,7 +76,7 @@ c9d318df7922bde3b265927b521ff5a965002826fc0cabfcaef52a56760f2d34 android.hardwar
 da13bd69282fb275767abb18704c57ff8038e6c139ad17157dc702810f70d06a android.hardware.gatekeeper@1.0::types
 37c7da4f823ec958dfa9c960e2d341c48f877e0bfa758f3fa9e2d9c1e1bd66d9 android.hardware.gnss@1.0::IAGnss
 7ec9afdb964bfb8369866913caf018f2636592885bcb558a65de2c5436ab4f60 android.hardware.gnss@1.0::IAGnssCallback
-d16e6a359be6963ea753d7138e84ecf2b93052097938938c4d36d7a47ea2e2ae android.hardware.gnss@1.0::IAGnssRil
+50746a16cc76f50cde00732260d5b9c8fe326d4beb5d2a0cf02cdc09307a4850 android.hardware.gnss@1.0::IAGnssRil
 2f907708d74d94b1e121ed27651c9c72af65952d347b58ff07dac5d5d7a7f678 android.hardware.gnss@1.0::IAGnssRilCallback
 5ac7edad06d76064b882be161f3f9d9692a997ec72e9f36addb7fe8918f49992 android.hardware.gnss@1.0::IGnss
 b05c983c87c3376e145223688c3b541b5e11b827f211e38d5a31af1ca3a2e222 android.hardware.gnss@1.0::IGnssBatching
diff --git a/gnss/1.0/IAGnssRil.hal b/gnss/1.0/IAGnssRil.hal
index 12289f6..cc820f8 100644
--- a/gnss/1.0/IAGnssRil.hal
+++ b/gnss/1.0/IAGnssRil.hal
@@ -68,14 +68,10 @@ interface IAGnssRil {
          */
         uint16_t lac;
 
+        uint16_t psc;
+
         /** Cell id in 2G. Utran Cell id in 3G. Cell Global Id EUTRA in LTE. */
         uint32_t cid;
-
-        /** Tracking Area Code in LTE. */
-        uint16_t tac;
-
-        /** Physical Cell id in LTE (not used in 2G and 3G) */
-        uint16_t pcid;
     };
 
     /** Represents ref locations */
diff --git a/gnss/1.0/default/AGnssRil.cpp b/gnss/1.0/default/AGnssRil.cpp
index 1458327..d97a1be 100644
--- a/gnss/1.0/default/AGnssRil.cpp
+++ b/gnss/1.0/default/AGnssRil.cpp
@@ -101,9 +101,8 @@ Return<void> AGnssRil::setRefLocation(const IAGnssRil::AGnssRefLocation& aGnssRe
         .mcc = cellID.mcc,
         .mnc = cellID.mnc,
         .lac = cellID.lac,
-        .cid = cellID.cid,
-        .tac = cellID.tac,
-        .pcid = cellID.pcid
+        .psc = cellID.psc,
+        .cid = cellID.cid
     };
 
     mAGnssRilIface->set_ref_location(&aGnssRefloc, sizeof(aGnssRefloc));
