diff --git a/Jelly/app/src/main/java/org/lineageos/jelly/MainActivity.java b/app/src/main/java/org/lineageos/jelly/MainActivity.java
index 71990dd..bc1e114 100644
--- a/Jelly/app/src/main/java/org/lineageos/jelly/MainActivity.java
+++ b/Jelly/app/src/main/java/org/lineageos/jelly/MainActivity.java
@@ -68,6 +68,7 @@ import android.view.WindowManager;
 import android.view.inputmethod.EditorInfo;
 import android.webkit.CookieManager;
 import android.webkit.URLUtil;
+import android.webkit.ValueCallback;
 import android.webkit.WebChromeClient;
 import android.widget.AutoCompleteTextView;
 import android.widget.FrameLayout;
@@ -131,6 +132,7 @@ public class MainActivity extends WebViewExtActivity implements
         }
     };
 
+    private ValueCallback<Uri[]> mPathCallback;
     private CoordinatorLayout mCoordinator;
     private AppBarLayout mAppBar;
     private FrameLayout mWebViewContainer;
@@ -708,6 +710,8 @@ public class MainActivity extends WebViewExtActivity implements
         mFullScreenCallback.onCustomViewHidden();
         mFullScreenCallback = null;
         mCustomView = null;
+
+        findViewById(R.id.app_bar_layout).requestFocus();
     }
 
     private void addShortcut() {
@@ -839,4 +843,16 @@ public class MainActivity extends WebViewExtActivity implements
             }
         }
     }
+
+    public void setPathCallback(ValueCallback<Uri[]> path) {
+        mPathCallback = path;
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent intent) {
+        if (requestCode == WebViewExtActivity.REQUEST_SELECT_FILE && mPathCallback != null) {
+            mPathCallback.onReceiveValue(WebChromeClient.FileChooserParams.parseResult(resultCode, intent));
+            mPathCallback = null;
+        }
+    }
 }
diff --git a/Jelly/app/src/main/java/org/lineageos/jelly/webview/ChromeClient.java b/app/src/main/java/org/lineageos/jelly/webview/ChromeClient.java
index acf91be..1a2fd46 100644
--- a/Jelly/app/src/main/java/org/lineageos/jelly/webview/ChromeClient.java
+++ b/Jelly/app/src/main/java/org/lineageos/jelly/webview/ChromeClient.java
@@ -76,9 +76,10 @@ class ChromeClient extends WebChromeClientCompat {
     @Override
     public boolean onShowFileChooser(WebView view, ValueCallback<Uri[]> path,
                                      FileChooserParams params) {
-        Intent intent = params.createIntent();
         try {
-            mActivity.startActivity(intent);
+            Intent intent = params.createIntent();
+            mActivity.setPathCallback(path);
+            mActivity.startActivityForResult(intent, WebViewExtActivity.REQUEST_SELECT_FILE);
         } catch (ActivityNotFoundException e) {
             Toast.makeText(mActivity, mActivity.getString(R.string.error_no_activity_found),
                     Toast.LENGTH_LONG).show();
diff --git a/Jelly/app/src/main/java/org/lineageos/jelly/webview/WebViewExtActivity.java b/app/src/main/java/org/lineageos/jelly/webview/WebViewExtActivity.java
index dcf02d4..0b90eb1 100644
--- a/Jelly/app/src/main/java/org/lineageos/jelly/webview/WebViewExtActivity.java
+++ b/Jelly/app/src/main/java/org/lineageos/jelly/webview/WebViewExtActivity.java
@@ -16,12 +16,16 @@
 package org.lineageos.jelly.webview;
 
 import android.graphics.Bitmap;
+import android.net.Uri;
 import android.support.v7.app.AppCompatActivity;
 import android.view.View;
+import android.webkit.ValueCallback;
 import android.webkit.WebChromeClient;
 
 public abstract class WebViewExtActivity extends AppCompatActivity {
 
+    public static final int REQUEST_SELECT_FILE = 100;
+
     public abstract void downloadFileAsk(String url, String contentDisposition, String mimeType);
 
     public abstract boolean hasLocationPermission();
@@ -37,4 +41,6 @@ public abstract class WebViewExtActivity extends AppCompatActivity {
     public abstract void onShowCustomView(View view, WebChromeClient.CustomViewCallback callback);
 
     public abstract void onHideCustomView();
+
+    public abstract void setPathCallback(ValueCallback<Uri[]> path);
 }
diff --git a/FMRadio/jni/fmr/Android.mk b/jni/fmr/Android.mk
index 2a86937..ce915cc 100644
--- a/FMRadio/jni/fmr/Android.mk
+++ b/FMRadio/jni/fmr/Android.mk
@@ -27,6 +27,7 @@ LOCAL_C_INCLUDES := $(JNI_H_INCLUDE) \
     frameworks/base/include/media
 
 LOCAL_SHARED_LIBRARIES := \
+    liblog \
     libcutils \
     libdl \
     libmedia \
