diff --git a/include/system/camera.h b/include/system/camera.h
index 0570ca0..81660a9 100644
--- a/include/system/camera.h
+++ b/include/system/camera.h
@@ -202,6 +202,10 @@ enum {
      */
 #ifndef CAMERA_VENDOR_L_COMPAT
     CAMERA_CMD_SET_VIDEO_FORMAT = 11,
+    CAMERA_CMD_SET_FLIP_ON = 12,
+    MULTI_FRAME_SHOT_START            = 1261,
+    AUTO_LOW_LIGHT_SET                = 1351,
+    HDR_PICTURE_MODE_CHANGE           = 1273,
 
     CAMERA_CMD_VENDOR_START = 20,
     /**
@@ -358,6 +362,11 @@ typedef struct camera_frame_metadata {
      * An array of the detected faces. The length is number_of_faces.
      */
     camera_face_t *faces;
+
+    /**
+     * Value for low light condition
+     */
+    int32_t light_condition;
 } camera_frame_metadata_t;
 
 __END_DECLS
diff --git a/init/builtins.cpp b/init/builtins.cpp
index 1681d17..f32dccc 100644
--- a/init/builtins.cpp
+++ b/init/builtins.cpp
@@ -920,6 +920,15 @@ int do_wait(int nargs, char **args)
         return -1;
 }
 
+int do_pipe(int nargs, char **args) {
+    mode_t mode = get_mode(args[1]);
+    if (mkfifo(args[2], mode) < 0) {
+        ERROR("peter do pipe error haha\n");
+    return -errno;
+    }
+    return 0;
+}
+
 /*
  * Callback to make a directory from the ext4 code
  */
diff --git a/init/init_parser.cpp b/init/init_parser.cpp
index c36d36e..0a6f1e5 100644
--- a/init/init_parser.cpp
+++ b/init/init_parser.cpp
@@ -173,6 +173,7 @@ static int lookup_keyword(const char *s)
         break;
     case 'p':
         if (!strcmp(s, "owerctl")) return K_powerctl;
+        if (!strcmp(s, "ipe")) return K_pipe; //sprd add for loopback test
         break;
     case 'r':
         if (!strcmp(s, "estart")) return K_restart;
diff --git a/init/keywords.h b/init/keywords.h
index 303685d..21521e5 100644
--- a/init/keywords.h
+++ b/init/keywords.h
@@ -39,6 +39,7 @@ int do_load_system_props(int nargs, char **args);
 int do_verity_load_state(int nargs, char **args);
 int do_verity_update_state(int nargs, char **args);
 int do_wait(int nargs, char **args);
+int do_pipe(int nargs, char **args);
 #define __MAKE_KEYWORD_ENUM__
 #define KEYWORD(symbol, flags, nargs, func) K_##symbol,
 enum {
@@ -100,6 +101,7 @@ enum {
     KEYWORD(verity_load_state,      COMMAND, 0, do_verity_load_state)
     KEYWORD(verity_update_state,    COMMAND, 0, do_verity_update_state)
     KEYWORD(wait,        COMMAND, 1, do_wait)
+    KEYWORD(pipe,        COMMAND, 2, do_pipe)
     KEYWORD(write,       COMMAND, 2, do_write)
     KEYWORD(writepid,    OPTION,  0, 0)
 #ifdef __MAKE_KEYWORD_ENUM__
diff --git a/libion/Android.mk b/libion/Android.mk
index 6562cd3..8263c7b 100644
--- a/libion/Android.mk
+++ b/libion/Android.mk
@@ -4,9 +4,10 @@ include $(CLEAR_VARS)
 LOCAL_SRC_FILES := ion.c
 LOCAL_MODULE := libion
 LOCAL_MODULE_TAGS := optional
-LOCAL_SHARED_LIBRARIES := liblog
-LOCAL_C_INCLUDES := $(LOCAL_PATH)/include $(LOCAL_PATH)/kernel-headers
-LOCAL_EXPORT_C_INCLUDE_DIRS := $(LOCAL_PATH)/include $(LOCAL_PATH)/kernel-headers
+LOCAL_SHARED_LIBRARIES := liblog libcutils
+LOCAL_C_INCLUDES := $(LOCAL_PATH)/include $(TARGET_OUT_INTERMEDIATES)/KERNEL_OBJ/usr/include/video
+LOCAL_EXPORT_C_INCLUDE_DIRS := $(LOCAL_PATH)/include $(TARGET_OUT_INTERMEDIATES)/KERNEL_OBJ/usr/include/video
+LOCAL_ADDITIONAL_DEPENDENCIES := $(TARGET_OUT_INTERMEDIATES)/KERNEL_OBJ/usr
 LOCAL_CFLAGS := -Werror
 include $(BUILD_SHARED_LIBRARY)
 
@@ -14,8 +15,9 @@ include $(CLEAR_VARS)
 LOCAL_SRC_FILES := ion.c ion_test.c
 LOCAL_MODULE := iontest
 LOCAL_MODULE_TAGS := optional tests
-LOCAL_C_INCLUDES := $(LOCAL_PATH)/include $(LOCAL_PATH)/kernel-headers
-LOCAL_SHARED_LIBRARIES := liblog
+LOCAL_C_INCLUDES := $(LOCAL_PATH)/include $(TARGET_OUT_INTERMEDIATES)/KERNEL_OBJ/usr/include/video
+LOCAL_ADDITIONAL_DEPENDENCIES := $(TARGET_OUT_INTERMEDIATES)/KERNEL_OBJ/usr
+LOCAL_SHARED_LIBRARIES := liblog libcutils
 LOCAL_CFLAGS := -Werror
 include $(BUILD_EXECUTABLE)
 
diff --git a/libion/include/ion/ion.h b/libion/include/ion/ion.h
index f47793d..d276b16 100644
--- a/libion/include/ion/ion.h
+++ b/libion/include/ion/ion.h
@@ -35,6 +35,7 @@ int ion_alloc(int fd, size_t len, size_t align, unsigned int heap_mask,
 int ion_alloc_fd(int fd, size_t len, size_t align, unsigned int heap_mask,
               unsigned int flags, int *handle_fd);
 int ion_sync_fd(int fd, int handle_fd);
+int ion_invalidate_fd(int fd, int handle_fd);
 int ion_free(int fd, ion_user_handle_t handle);
 int ion_map(int fd, ion_user_handle_t handle, size_t length, int prot,
             int flags, off_t offset, unsigned char **ptr, int *map_fd);
diff --git a/libion/ion.c b/libion/ion.c
index d1984bd..a411c76 100644
--- a/libion/ion.c
+++ b/libion/ion.c
@@ -31,6 +31,11 @@
 #include <linux/ion.h>
 #include <ion/ion.h>
 
+struct file;
+
+#include <stdbool.h>
+#include "ion_sprd.h"
+
 int ion_open()
 {
     int fd = open("/dev/ion", O_RDWR);
@@ -169,6 +174,16 @@ int ion_import(int fd, int share_fd, ion_user_handle_t *handle)
     return ret;
 }
 
+int ion_invalidate_fd(int fd, int handle_fd)
+{
+	struct ion_custom_data custom_data;
+	if (handle_fd < 0)
+		return -EINVAL;
+	custom_data.cmd = ION_SPRD_CUSTOM_INVALIDATE;
+	custom_data.arg = (unsigned long)handle_fd;
+	return ioctl(fd, ION_IOC_CUSTOM, &custom_data);
+}
+
 int ion_sync_fd(int fd, int handle_fd)
 {
     struct ion_fd_data data = {
